zcash_client_sqlite/src/wallet.rs

```diff
@@ -2048,4 +2049,138 @@ mod tests {
+                    .unwrap()
+                    .get(taddr)
+                    .cloned()
+                    .unwrap_or(Amount::zero()),
+                Amount::from(expected),
+            );
+            assert_eq!(
+                st.wallet()
+                    .get_unspent_transparent_outputs(taddr, max_height, &[])
+                    .unwrap()
+                    .into_iter()
+                    .map(|utxo| utxo.value())
+                    .sum::<Option<Amount>>(),
+                Some(Amount::from(expected)),
+            );
+        };
+
+        // The wallet starts out with zero balance.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+
+        // Create a fake transparent output.
+        let value = NonNegativeAmount::from_u64(100000).unwrap();
+        let outpoint = OutPoint::new([1u8; 32], 1);
+        let txout = TxOut {
+            value: value.into(),
+            script_pubkey: taddr.script(),
+        };
+
+        // Pretend the output was received in the chain tip.
+        let height = st.wallet().chain_height().unwrap().unwrap();
+        let utxo =
+            WalletTransparentOutput::from_parts(outpoint.clone(), txout.clone(), height).unwrap();
+        st.wallet_mut()
+            .put_received_transparent_utxo(&utxo)
+            .unwrap();
+
+        // The wallet should detect the balance as having 1 confirmation.
+        check_balance(&st, 0, value);
+        check_balance(&st, 1, value);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Shield the output.
+        let input_selector = GreedyInputSelector::new(
+            fixed::SingleOutputChangeStrategy::new(FixedFeeRule::non_standard(Amount::zero())),
+            DustOutputPolicy::default(),
+        );
+        let txid = st
+            .shield_transparent_funds(
+                &input_selector,
+                value,
+                &usk,
+                &[*taddr],
+                &MemoBytes::empty(),
+                NonZeroU32::new(1).unwrap(),
+            )
+            .unwrap();
+
+        // The wallet should have zero transparent balance, because the shielding
+        // transaction can be mined.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Mine the shielding transaction.
+        let (mined_height, _) = st.generate_next_block_including(txid);
+        st.scan_cached_blocks(mined_height, 1);
+
+        // The wallet should still have zero transparent balance.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Unmine the shielding transaction via a reorg.
+        st.wallet_mut()
+            .truncate_to_height(mined_height - 1)
+            .unwrap();
```

Comment:
Hm, yeah, that `Verify` range insertion was from pretty early in my implementation, and I don't think it necessarily makes sense any more. Instead, after the rewind the next `update_chain_tip` call should establish the scanning ranges.

Reply:
I think there is still a benefit to having a `Verify` range at the post-truncation tip, because the caller may not have rewound far enough, and if the new chain tip is scanned before we check the connection then the subsequent truncation will undo that new scan as well (causing more duplication of scanning work).

But it should be a verify range that doesn't extend past the truncated tip, which means we need at least one block of overlap (unlike the `Verify` range that `update_chain_tip` creates which sits adjacent to the prior tip, sufficient for connectivity checking but avoiding duplicate scanning). Probably just adding a 1-block range rescanning the truncated tip is sufficient; we already tell the caller to pick some rewind distance themselves, so there's no need to bake in additional duplicate scanning unless we want to change that interaction pattern (e.g. by having APIs that tell the caller what height to rewind to).

Reply:
What if the `Verify` range was just one block long?

Reply:
Hmm, yeah. It's probably reasonable to assume here that the caller rewound to a block before a problematic scan, which implies that there exists at least one block after the truncation height. So instead of my current fix (verify the truncation height), we could verify the block after the truncation height (but no further). Still slightly unintuitive in terms of the method name though, as `chain_height` would not match the truncation.

Reply:
I'm going to leave my current fix in place (rescan the truncated-to height, so the chain height doesn't differ from that height until `update_chain_tip` is called) unless review asks otherwise.